### ISSUE TITLE
RUF009 - Frozen Dataclass default should be valid

### ISSUE THREAD

**Issue by CarliJoy** [astral-sh/ruff#17424]
### Summary

Frozen dataclasses and attrs objects should be considered immutable.

[Playground](https://play.ruff.rs/b092eb6e-c17e-47dc-b45c-fdba55e00852)

```python
@attrs.frozen
class A:
    foo: int = 1

@attrs.frozen
class B:
    a: A = A()  # should be valid, but gives RUF009

@dataclasses.dataclass(frozen=True)
class C:
    foo: int = 1

@dataclasses.dataclass(frozen=True)
class D:
    d: C = C()  # should be valid , but gives RUF009
```

### Version

v0.11.5

**Comment by dhruvmanila** [comment]
I don't think that's what the rule is intended to catch. When you specify `frozen = True`, it's only the dataclass fields that are frozen i.e., you cannot re-assign the fields, and not the objects contained within the dataclass. Consider the following example:
```pycon
>>> from dataclasses import dataclass
>>> class Foo:
...     def __init__(self, x: int) -> None:
...         self.x = x
...         
>>> @dataclass(frozen=True)
... class Bar:
...     f: Foo = Foo(1)
...     
>>> b1 = Bar()
>>> b2 = Bar()
>>> b3 = Bar()
>>> b1.f.x
1
>>> b2.f.x
1
>>> b3.f.x
1
>>> b1.f.x = 2
>>> b1.f.x
2
>>> b2.f.x
2
>>> b3.f.x
2
>>> b1.f = Foo(2)
Traceback (most recent call last):
  File "<python-input-13>", line 1, in <module>
    b1.f = Foo(2)
    ^^^^
  File "<string>", line 15, in __setattr__
dataclasses.FrozenInstanceError: cannot assign to field 'f'
```

You can see that updating the `Foo` object's `x` attribute on one instance of `Bar` (which is a dataclass) is visible to all instances of the same dataclass.

**Comment by dhruvmanila** [comment]
And, the way to fix that would be to use `field(default_factory=...)`:

```pycon
>>> from dataclasses import field
>>> @dataclass(frozen=True)
... class FixedBar:
...     f: Foo = field(default_factory=lambda: Foo(1))
...     
>>> fb1 = FixedBar()
>>> fb2 = FixedBar()
>>> fb1.f.x
1
>>> fb2.f.x
1
>>> fb1.f.x = 2
>>> fb1.f.x
2
>>> fb2.f.x
1
```

**Comment by dhruvmanila** [comment]
Oh shoot, I think I misunderstood the issue 🤦‍♂

**Comment by lubaskinc0de** [comment]
I would like to start working on this from Monday.
